crypto/provider_core.c

```diff
@@ -550,6 +560,7 @@ static int provider_init(OSSL_PROVIDER *prov)
 
             OPENSSL_free(merged_path);
             OPENSSL_free(allocated_path);
+            OPENSSL_free((char*)load_dir);
```

Comment:
This cast actually fairly well indicates you have a problem here. I'd suggest using separate allocated_load_dir variable to handle the conditionally allocated value and assign it above from the OPENSSL_strdup().

Reply:
Definitely, good catch.